test/hotspot/jtreg/compiler/lib/compile_framework/Utils.java

```diff
@@ -0,0 +1,171 @@
+                if (!asmtools.exists()) {
+                    throw new InternalCompileFrameworkException("Found jtreg.jar in classpath, but could not find asmtools.jar");
+                }
+                return asmtools.getAbsolutePath();
+            }
+        }
+        throw new InternalCompileFrameworkException("Could not find asmtools because could not find jtreg.jar in classpath");
+    }
+
+    private static void writeCodeToFile(String code, Path path) {
+        printlnVerbose("File: " + path);
+
+        // Ensure directory of the file exists.
+        Path dir = path.getParent();
+        try {
+            Files.createDirectories(dir);
+        } catch (Exception e) {
+            throw new CompileFrameworkException("Could not create directory: " + dir, e);
+        }
+
+        // Write to file.
+        try (BufferedWriter writer = Files.newBufferedWriter(path)) {
+            writer.write(code);
+        } catch (Exception e) {
+            throw new CompileFrameworkException("Could not write file: " + path, e);
+        }
+    }
+
+    /**
+    * Write sources to file.
+    */
+    public static List<Path> writeSourcesToFile(List<SourceCode> sources, Path sourceDir) {
+        List<Path> storedFiles = new ArrayList<Path>();
+        for (SourceCode sourceCode : sources) {
+            Path path = sourceDir.resolve(sourceCode.filePathName());
+            writeCodeToFile(sourceCode.code(), path);
+            storedFiles.add(path);
+        }
+        return storedFiles;
+    }
+
+    /**
+    * Execute a given compilation, given as a {@code command}.
+    */
+    public static void executeCompileCommand(List<String> command) {
+        printlnVerbose("Compile command: " + String.join(" ", command));
+
+        ProcessBuilder builder = new ProcessBuilder(command);
+        builder.redirectErrorStream(true);
+
+        String output;
+        int exitCode;
+        try {
+            Process process = builder.start();
+            boolean exited = process.waitFor(COMPILE_TIMEOUT, TimeUnit.SECONDS);
+            if (!exited) {
+                process.destroyForcibly();
+                System.out.println("Timeout: compile command: " + String.join(" ", command));
+                throw new InternalCompileFrameworkException("Process timeout: compilation took too long.");
+            }
+            output = new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
+            exitCode = process.exitValue();
+        } catch (IOException e) {
+            throw new InternalCompileFrameworkException("IOException during compilation", e);
+        } catch (InterruptedException e) {
+            throw new CompileFrameworkException("InterruptedException during compilation", e);
+        }
+
+        if (exitCode != 0 || !output.isEmpty()) {
+            System.out.println("Compilation failed.");
+            System.out.println("Exit code: " + exitCode);
+            System.out.println("Output: '" + output + "'");
```

Comment:
You could print to System.err:
```suggestion
            System.err.println("Compilation failed.");
            System.err.println("Exit code: " + exitCode);
            System.err.println("Output: '" + output + "'");
```